src/connector/src/parser/scalar_adapter.rs

```diff
@@ -0,0 +1,217 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::str::FromStr;
+
+use anyhow::anyhow;
+use bytes::BytesMut;
+use pg_bigdecimal::PgNumeric;
+use risingwave_common::types::{DataType, Int256, ScalarImpl, ScalarRefImpl};
+use thiserror_ext::AsReport;
+use tokio_postgres::types::{to_sql_checked, FromSql, IsNull, Kind, ToSql, Type};
+
+use crate::error::ConnectorResult;
+
+#[derive(Clone, Debug)]
+pub struct EnumString(pub String);
+
+impl<'a> FromSql<'a> for EnumString {
+    fn from_sql(
+        _ty: &Type,
+        raw: &'a [u8],
+    ) -> Result<Self, Box<dyn std::error::Error + 'static + Sync + Send>> {
+        Ok(EnumString(String::from_utf8_lossy(raw).into_owned()))
+    }
+
+    fn accepts(ty: &Type) -> bool {
+        matches!(ty.kind(), Kind::Enum(_))
+    }
+}
+
+impl ToSql for EnumString {
+    to_sql_checked!();
+
+    fn accepts(ty: &Type) -> bool {
+        matches!(ty.kind(), Kind::Enum(_))
+    }
+
+    fn to_sql(
+        &self,
+        ty: &Type,
+        out: &mut BytesMut,
+    ) -> Result<IsNull, Box<dyn std::error::Error + Sync + Send>>
+    where
+        Self: Sized,
+    {
+        match ty.kind() {
+            Kind::Enum(e) => {
+                if e.contains(&self.0) {
+                    out.extend_from_slice(self.0.as_bytes());
+                    Ok(IsNull::No)
+                } else {
+                    Err(format!(
+                        "EnumString value {} is not in the enum type {:?}",
+                        self.0, e
+                    )
+                    .into())
+                }
+            }
+            _ => Err("EnumString can only be used with ENUM types".into()),
+        }
+    }
+}
+
+/// Adapter for `ScalarImpl` to Postgres data type,
+/// which can be used to encode/decode to/from Postgres value.
+#[derive(Debug)]
+pub(crate) enum ScalarAdapter<'a> {
+    Builtin(ScalarRefImpl<'a>),
+    Uuid(uuid::Uuid),
+    Numeric(PgNumeric),
+    Enum(EnumString),
+}
+
+impl ToSql for ScalarAdapter<'_> {
+    to_sql_checked!();
+
+    fn to_sql(
+        &self,
+        ty: &Type,
+        out: &mut bytes::BytesMut,
+    ) -> Result<IsNull, Box<dyn std::error::Error + Sync + Send>> {
+        match self {
+            ScalarAdapter::Builtin(v) => v.to_sql(ty, out),
+            ScalarAdapter::Uuid(v) => v.to_sql(ty, out),
+            ScalarAdapter::Numeric(v) => v.to_sql(ty, out),
+            ScalarAdapter::Enum(v) => v.to_sql(ty, out),
+        }
+    }
+
+    fn accepts(_ty: &Type) -> bool {
+        true
+    }
+}
+
+/// convert from Postgres uuid, numeric and enum to `ScalarAdapter`
+impl<'a> FromSql<'a> for ScalarAdapter<'_> {
+    fn from_sql(
+        ty: &Type,
+        raw: &'a [u8],
+    ) -> Result<Self, Box<dyn std::error::Error + Sync + Send>> {
+        match ty.kind() {
+            Kind::Simple => match *ty {
```

Comment:
Let's put a comment here:
> For array types (`Kind::Array`), `ScalarAdapter` is only meant to work as its elements rather than the whole array. The caller shall use `Option<Vec<Option<ScalarAdapter>>>` rather than `Option<ScalarAdapter>`.